### PR TITLE
[Feature] Allow `!ref` and `!flatten` as well as defining constants

### PR DESCRIPTION
**Pull request by grahamjenson**

### CODE REVIEW

File: lib/geoengineer/gps/constants.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Constants contains the needed information to
```
Comment (brad-larson) [inline lib/geoengineer/gps/constants.rb]: unfinished comment

File: lib/geoengineer/gps/constants.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @base_hash = base_hash
+
+    # _defaults is copied into each environment
+    @defaults_hash = @base_hash["_defaults"]
+
+    # remove all _'s
+    @constants_hash = HashUtils.remove_(@base_hash)
+
+    # the local environment overrides the values
+    @constants_hash.each_pair do |environment, vals|
```
Comment (brad-larson) [inline lib/geoengineer/gps/constants.rb]: Could do a `@constants_hash = HashUtils.deep_merge(@defaults_hash, @constants_hash)`

File: lib/geoengineer/gps/yaml_tag.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  end
+
+  def initialize(type, value, &block)
+    @type = type
+    @value = value
+    @block = block
+  end
+
+  def environment=(environment)
+    raise "Cannot change Tag Environemtn from #{@environment} to #{environment}" if @environment
```
Comment (brad-larson) [inline lib/geoengineer/gps/yaml_tag.rb]: spelling

File: lib/geoengineer/gps/constants.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      @constants_hash[environment] = @defaults_hash.merge(vals)
+    end
+
+    # Force different Tag objects
+    @constants_hash = HashUtils.deep_dup(@constants_hash)
+
+    # attach constants and environment to TTags
```
Comment (brad-larson) [inline lib/geoengineer/gps/constants.rb]: single T

File: lib/geoengineer/gps/constants.rb (hunk 2 above)
Comment (grahamjenson) [reply lib/geoengineer/gps/constants.rb]: not sure we want a deep merge. This is more for an override of top level keys

Comment (grahamjenson) [reply lib/geoengineer/gps/constants.rb]: e.g. sometime you want to remove all values to make an environment equal an empty hash.

File: lib/geoengineer/gps/yaml_tag.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  end
+end
+
+# Ref takes a query as input and replaces it with the value
+YAML.add_domain_type("", "ref") do |type, reference|
+  # If a string starts with `:` in ruby it treats it as a symbol
+  # to make references we add back a `:` to the string
+  reference = ":#{reference}" if reference.is_a?(Symbol)
+  GeoEngineer::GPS::YamlTag.new(type, reference) do |value|
+    finder.dereference!(reference).to_json
```
Comment (brad-larson) [inline lib/geoengineer/gps/yaml_tag.rb]: why not just `dereference.first`?  I find the `dereference!` vs `dereference` a bit arbitrary and mildly confusing

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-25 20:03:08 UTC
User failed mfa authentication

File: lib/geoengineer/gps/yaml_tag.rb (hunk 5 above)
Comment (grahamjenson) [reply lib/geoengineer/gps/yaml_tag.rb]: dereference! throws an error if it cant find the value or it finds more than one value.

File: lib/geoengineer/gps/gps.rb [hunk 6]
```diff
@@ -142,60 +152,54 @@ def expanded_hash
...
     end
   end
 
-  def expand_meta_nodes(projects_hash)
-    # We dup the original hash because we cannot edit and loop over it at the same time
-    loop_projects_hash(HashUtils.deep_dup(projects_hash)) do |node|
-      next unless node.meta?
+  def expand_meta_node(node, all_nodes)
+    expanded_nodes = []
+    bn = node.build_nodes
```
Comment (brad-larson) [inline lib/geoengineer/gps/gps.rb]: nit: this is a bad variable name

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-26 00:10:33 UTC
User failed mfa authentication

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-26 00:10:42 UTC
User failed mfa authentication

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-26 00:12:09 UTC
User failed mfa authentication

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-26 00:12:19 UTC
User failed mfa authentication

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-26 00:13:53 UTC
User failed mfa authentication

**Comment by heimdall-asguard** [comment]
Review Error for brad-larson @ 2019-02-26 00:14:03 UTC
User failed mfa authentication

**Pull request merged**